File: tests/test_plugin_trust.py
"""Parity: dev_server.py canonical plugin hash vs the shared vector.
Twin of tests/test_plugin_trust.js (Node) — both must agree, guaranteeing the
browser and the server compute the same identity hash over the same bytes."""
import base64
import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
import dev_server as ds

v = json.loads((Path(__file__).parent / "plugin-trust-vector.json").read_text(encoding="utf-8"))
fails = 0

for c in v["files"]:
    got = ds._sha256_bytes(base64.b64decode(c["b64"]))
    if got != c["fileHash"]:
        fails += 1
        print(f"  FAIL fileHash {c['rel']}: {got} != {c['fileHash']}")

composite = ds._plugin_hash({c["rel"]: c["fileHash"] for c in v["files"]})
if composite != v["pluginHash"]:
    fails += 1
    print(f"  FAIL pluginHash: {composite} != {v['pluginHash']}")

sf = v["singleFile"]
if ds._plugin_hash({sf["rel"]: sf["fileHash"]}) != sf["pluginHash"]:
    fails += 1
    print("  FAIL singleFile pluginHash")

if fails:
    print(f"{fails} PARITY FAILURES")
    sys.exit(1)
print(f"ALL {len(v['files']) + 2} PLUGIN-TRUST HASH CASES PASSED (python)")

File: tests/test_plugin_trust.js
/* Parity: js/core/plugin-trust.js canonical hash vs the shared vector (Node).
   Twin of tests/test_plugin_trust.py — both must agree. */
'use strict';
const fs = require('fs');
const path = require('path');

// Node exposes WebCrypto as globalThis.crypto (>=20). plugin-trust.js reads crypto.subtle.
if (typeof crypto === 'undefined') global.crypto = require('crypto').webcrypto;
const PluginTrust = require(path.join(__dirname, '..', 'js', 'core', 'plugin-trust.js'));

const v = JSON.parse(fs.readFileSync(path.join(__dirname, 'plugin-trust-vector.json'), 'utf-8'));

(async () => {
  let fails = 0;
  for (const c of v.files) {
    const bytes = new Uint8Array(Buffer.from(c.b64, 'base64'));
    const got = await PluginTrust.fileHash(bytes);
    if (got !== c.fileHash) { fails++; console.log(`  FAIL fileHash ${c.rel}: ${got} != ${c.fileHash}`); }
  }
  const fhs = {};
  for (const c of v.files) fhs[c.rel] = c.fileHash;
  const composite = await PluginTrust.pluginHash(fhs);
  if (composite !== v.pluginHash) { fails++; console.log(`  FAIL pluginHash: ${composite} != ${v.pluginHash}`); }

  const sf = v.singleFile;
  const single = await PluginTrust.pluginHash({ [sf.rel]: sf.fileHash });
  if (single !== sf.pluginHash) { fails++; console.log('  FAIL singleFile pluginHash'); }

  if (fails) { console.log(`${fails} PARITY FAILURES`); process.exit(1); }
  console.log(`ALL ${v.files.length + 2} PLUGIN-TRUST HASH CASES PASSED (js)`);
})();
